plugins/inputs/s7comm/s7comm.go

```diff
@@ -0,0 +1,402 @@
+
+type metricDefinition struct {
+	Name   string                  `toml:"name"`
+	Fields []metricFieldDefinition `toml:"fields"`
+	Tags   map[string]string       `toml:"tags"`
+}
+
+type converterFunc func([]byte) interface{}
+
+type batch struct {
+	items    []gos7.S7DataItem
+	mappings []fieldMapping
+}
+
+type fieldMapping struct {
+	measurement string
+	field       string
+	tags        map[string]string
+	convert     converterFunc
+}
+
+// S7comm represents the plugin
+type S7comm struct {
+	Server          string             `toml:"server"`
+	Rack            int                `toml:"rack"`
+	Slot            int                `toml:"slot"`
+	Timeout         config.Duration    `toml:"timeout"`
+	DebugConnection bool               `toml:"debug_connection"`
+	Configs         []metricDefinition `toml:"metric"`
+	Log             telegraf.Logger    `toml:"-"`
+
+	handler *gos7.TCPClientHandler
+	client  gos7.Client
+	batches []batch
+}
+
+// SampleConfig returns a basic configuration for the plugin
+func (*S7comm) SampleConfig() string {
+	return sampleConfig
+}
+
+// Init checks the config settings and prepares the plugin. It's called
+// once by the Telegraf agent after parsing the config settings.
+func (s *S7comm) Init() error {
+	// Check settings
+	if s.Server == "" {
+		return errors.New("'server' has to be specified")
+	}
+	if s.Rack < 0 {
+		return errors.New("'rack' has to be specified")
+	}
+	if s.Slot < 0 {
+		return errors.New("'slot' has to be specified")
+	}
+
+	if len(s.Configs) == 0 {
+		return errors.New("no metric defined")
+	}
+
+	// Create the requests
+	return s.createRequests()
+}
+
+// Start initializes the connection to the remote endpoint
+func (s *S7comm) Start(_ telegraf.Accumulator) error {
+	// Create handler for the connection
+	s.handler = gos7.NewTCPClientHandler(s.Server, s.Rack, s.Slot)
+	s.handler.Timeout = time.Duration(s.Timeout)
+	if s.DebugConnection {
+		s.handler.Logger = log.New(os.Stderr, "D! [inputs.s7comm]", log.LstdFlags)
+	}
+	if err := s.handler.Connect(); err != nil {
+		return fmt.Errorf("connecting to %q failed: %w", s.Server, err)
+	}
+	s.client = gos7.NewClient(s.handler)
+
+	return nil
+}
+
+// Stop disconnects from the remote endpoint and cleans up
+func (s *S7comm) Stop() {
+	if s.handler != nil {
+		s.handler.Close()
+	}
+}
+
+// Gather collects the data from the device
+func (s *S7comm) Gather(acc telegraf.Accumulator) error {
+	timestamp := time.Now()
+	grouper := metric.NewSeriesGrouper()
+
+	for i, b := range s.batches {
+		// Read the batch
+		s.Log.Debugf("Reading batch %d...", i+1)
+		if err := s.client.AGReadMulti(b.items, len(b.items)); err != nil {
+			return fmt.Errorf("reading batch %d failed: %w", i+1, err)
+		}
+
+		// Dissect the received data into fields
+		for j, m := range b.mappings {
+			// Convert the data
+			buf := b.items[j].Data
+			value := m.convert(buf)
+			s.Log.Debugf("  got %v for field %q @ %d (%T)--> %v (%T)", buf, m.field, b.items[j].Start, value, value)
```

Comment:
Is there a reason for the extra whitespace? Also are there 5 values or 6? Does one item return an extra?

Reply:
Yeah the whitespace helps me in debug to align the fields to the requests (same in modbus IIRC). But the arguments are a good catch... Will fix. Do you also want me to remove the indentation?

Reply:
indent is fine, just double checking it was done on purpose